### ISSUE TITLE
SRS-X88 status/output

### ISSUE THREAD

**Issue by seanb-uk** [rytilahti/python-songpal#3]
On to the next set of issues with the SRS-X88:
```
songpal --endpoint http://192.168.1.18:54480/sony status
Power on
(output truncated)
  File "/home/sean/python3/lib/python3.5/site-packages/songpal/main.py", line 103, in status
    vol = await dev.get_volume_information()
  File "/home/sean/python3/lib/python3.5/site-packages/songpal/protocol.py", line 254, in get_volume_information
    volume_info = [Volume(self.services, **x) for x in res]
  File "/home/sean/python3/lib/python3.5/site-packages/songpal/protocol.py", line 254, in <listcomp>
    volume_info = [Volume(self.services, **x) for x in res]
TypeError: __init__() missing 1 required positional argument: 'output'
```
Calling output on its own:

```
songpal --endpoint http://192.168.1.18:54480/sony output
Traceback (most recent call last):
(ouput truncated)
  File "/home/sean/python3/lib/python3.5/site-packages/songpal/protocol.py", line 176, in get_outputs
    return [Output(self.services, **x) for x in res]
  File "/home/sean/python3/lib/python3.5/site-packages/songpal/protocol.py", line 176, in <listcomp>
    return [Output(self.services, **x) for x in res]
TypeError: __init__() missing 3 required positional arguments: 'active', 'label', and 'iconUrl'
```
When I run it with debug on I get the following before the traceback:
```
DEBUG:songpal.method:got res: {'id': 12,
 'result': [[{'connection': 'unknown',
              'meta': 'meta:usbdac',
              'title': 'USB DAC',
              'uri': 'extInput:usbDac'},
             {'connection': 'unknown',
              'meta': 'meta:btaudio',
              'title': 'Bluetooth Audio',
              'uri': 'extInput:btAudio'},
             {'connection': 'unknown',
              'meta': 'meta:line',
              'title': 'Audio in',
              'uri': 'extInput:line?port=1'}]]}
```

So it looks like it's finding a partial list but not indicating which is currently selected.

Incidentally is this actually the set of inputs (i.e. sources) rather than outputs? If so I'd expect to see the network inputs in the list - the device supports Chromecast and Spotify.

Let me know if you want a full dump_devinfo or full debug output for any of the commands or if there are any other commands you'd like me to try.

**Comment by rytilahti** [comment]
For volume, you can type:
```
songpal command audio getVolumeInformation '{}'
```
to see what kind of output it provides. I'm starting to rethink if it was a good idea to use attrs library for those containers considering how different the outputs are (the changes can be related either to different API versions or implementations..).

For output the command is `songpal command avContent getCurrentExternalTerminalsStatus`, and they are really inputs indeed. I will change it later to reflect the reality. I'm unsure how chromecast or spotify are visible over this API if at all, I think chromecast-support uses chromecast protocol for comms (see https://github.com/balloob/pychromecast).

A full `dump_devinfo` output would be nice to have, if you want to upload it here you can also e-mail it to me. I'll have to think abut what commands should be executed to get a baseline for features to support in the client tool.

**Comment by seanb-uk** [comment]
Unfortunately dump_devinfo also throws an error:

`TypeError: __init__() got an unexpected keyword argument 'ssid'`

The list of outputs is what I get in the message above, running output. It shows the first three but there are five on the remote: bluetooth, line in, USB A, USB B and network. The device works with pychromecast when the input is set to network.

For my uses I'd just like to be able to turn it on, change the input and adjust the volume level. 

I've attached my dump_devinfo run with debug on (deleted the ssid and mac address values):

[srsx88_dump_devinfo.txt](https://github.com/rytilahti/python-songpal/files/1584742/srsx88_dump_devinfo.txt)

**Comment by rytilahti** [comment]
Looking into that partial devinfo dump, it looks like the device reports that method to respond with those missing labels. This may be caused by an improper version information handling (could you please upload a new dump_devinfo with `-dd` to make it also output the raw payloads.

I added those keys temporarily as non-mandatory, so if you update your checkout it should not raise an exception anymore (for that ssid, nor for those you mentioned earlier). This is just a temporary though, as not having the `active` for an "output"/"input" does not allow checking what is currently active.

**Comment by seanb-uk** [comment]
It still gave me the ssid error at first, but I changed line 112 in containers.py from:
`    bssid = attr.ib(default=None)`
To
`    ssid = attr.ib(default=None)`

And it worked, so guess it was a typo?

New dump_devinfo with -dd attached.

I've done my own digging and found that the current input is contained in the source attribute returned by getPlayingContentInfo:

```
songpal --endpoint http://192.168.1.18:54480/sony command avContent getPlayingContentInfo {""}
Calling avContent.getPlayingContentInfo with params {}
INFO:songpal.protocol:Calling avContent.getPlayingContentInfo({})
[{'contentKind': 'input', 'uri': 'extInput:btAudio', 'stateInfo': {'state': ''}, 'parentUri': '', 'source': 'extInput:btAudio'}]
```
This can take values not listed in getCurrentTerminalsStatus - for example, when on network input:

```
Calling avContent.getPlayingContentInfo with params {}
INFO:songpal.protocol:Calling avContent.getPlayingContentInfo({})
[{'stateInfo': {'supplement': '', 'state': 'STOPPED'}, 'uri': 'netService:audio?service=spotify&contentId=-1', 'contentKind': 'service', 'service': 'netService:audio?service=spotify', 'source': 'netService:audio', 'fileNo': '255', 'parentUri': ''}]
```
I'm still looking to see if I can find a call that lists all of the sources.

[srsx88_dump_devinfo2.txt](https://github.com/rytilahti/python-songpal/files/1592937/srsx88_dump_devinfo2.txt)

**Comment by rytilahti** [comment]
I just released 0.0.3 which has a more resilient way for parsing the input from devices. You should see (even when not in debug mode) a warning about extra information not currently handled, so please create new issues based on that information :-)
